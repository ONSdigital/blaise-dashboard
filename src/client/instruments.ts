import { Instrument } from "blaise-api-node-client";
import axios from "axios";

export async function getInstruments(): Promise<Instrument[]> {
  const response = await axios.get("/api/instruments");

  return response.data;
}
