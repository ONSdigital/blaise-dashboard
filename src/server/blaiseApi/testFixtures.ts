import {Outcome} from "blaise-api-node-client";

export const mockQuestionnaireList = [
        {
            name: "LMS2101_CC1",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        },
        {
            name: "OPN2101A",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        },
        {
            name: "OPN2007T",
            installDate: "210122",
            serverParkName: "gusty",
            fieldPeriod: "210122"
        }];

export const mockCaseList = [
            {
                primaryKey: "1000001" ,
                outcome: Outcome.Completed
            },
            {
                primaryKey: "1000002" ,
                outcome: Outcome.Completed
            },
            {
                primaryKey: "1000003" ,
                outcome: Outcome.None
            },
            {
                primaryKey: "1000004" ,
                outcome: Outcome.Partial
            },
            {
                primaryKey: "1000005" ,
                outcome: Outcome.AppointmentMade
            },
            {
                primaryKey: "1000006" ,
                outcome: Outcome.NonContact
            },
            {
                primaryKey: "1000007" ,
                outcome: 999
            }];