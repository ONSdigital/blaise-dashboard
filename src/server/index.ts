import BlaiseApiClient from "blaise-api-node-client";
import { GetConfigFromEnv } from "./config";
import NewServer from "./server";

const port: string = process.env.PORT || "5000";
const config = GetConfigFromEnv();
const blaiseApiClient = new BlaiseApiClient(config.BlaiseApiUrl);
const app = NewServer(blaiseApiClient, config);

app.listen(port);

console.log("App is listening on port " + port);