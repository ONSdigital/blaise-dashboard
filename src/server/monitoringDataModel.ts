export type MonitoringDataModel = {
    hostname: string,
    regions: Region[]
};

export type Region = {
    region: string,
    status: string
}

export type UptimeCheck = {
    hostname: string,
    eurBelgium: string , 
    apacSingapore: string,
    northAmerica: string,
    southAmerica: string
}