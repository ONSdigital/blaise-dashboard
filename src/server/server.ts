import express, { Request, Response, Express } from "express";
import path from "path";
import ejs from "ejs";
import HealthCheckHandler from "./handlers/healthCheckHandler";
import QuestionnaireListHandler from "./handlers/questionnaireListHandler";
import { Config } from "./config";
import BlaiseApiClient from "blaise-api-node-client";
import caseReportHandler from "./handlers/reportHandler";
import NodeCache from "node-cache";
import monitoringHandler from "./handlers/monitoringHandler";

function NewServer(blaiseApiClient: BlaiseApiClient, cache: NodeCache, config: Config): Express {
    const server = express();
    const buildFolder = "../build";
    server.set("views", path.join(__dirname, buildFolder));
    server.engine("html", ejs.renderFile);
    server.use("/static", express.static(path.join(__dirname, `${buildFolder}/static`)));

    server.use(express.json());
    server.use(express.urlencoded({ extended: true }));

    //define handlers
    server.use("/", HealthCheckHandler());
    server.use("/", QuestionnaireListHandler(blaiseApiClient, cache, config));
    server.use("/", caseReportHandler(blaiseApiClient, cache, config));
    server.use("/", monitoringHandler());

    //define entry point
    server.get("*", function (req: Request, res: Response) {
        res.render("index.html");
    });

    server.use(function (err: Error, req: Request, res: Response) {
        res.render("../src/views/500.html", {});
    });
    return server;
}

export default NewServer;
