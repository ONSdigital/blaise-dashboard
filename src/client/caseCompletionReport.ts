import { CaseCompletionReport } from "../server/blaiseApi/caseCompletionReport";
import axios from "axios";

export async function getCaseCompletionReport(questionnaireName: string): Promise<CaseCompletionReport> {
  const response = await axios.get(`/api/reports/cases/completions/${questionnaireName}`);

  return response.data;
}
