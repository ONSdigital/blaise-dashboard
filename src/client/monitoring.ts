import axios from "axios";
import { MonitoringDataModel } from "../server/monitoringDataModel";

export async function getMonitoring(): Promise<MonitoringDataModel[]> {
  const response = await axios.get("/api/monitoring");
  return response.data;
}
