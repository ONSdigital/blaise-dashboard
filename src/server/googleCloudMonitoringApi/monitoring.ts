import {MetricServiceClient, UptimeCheckServiceClient} from "@google-cloud/monitoring";
import {MonitoringDataModel, Region} from "../monitoringDataModel";
import {google} from "@google-cloud/monitoring/build/protos/protos";
import * as protos from "@google-cloud/monitoring/build/protos/protos";

const regionsMonitored: string [] = ["eur-belgium", "apac-singapore", "usa-oregon", "sa-brazil-sao_paulo"];

export type GetUptimeChecksConfigResult = protos.google.monitoring.v3.IUptimeCheckConfig[];

export type ListTimeSeriesResult = protos.google.monitoring.v3.ITimeSeries[];

export interface GoogleMonitoring {
    getUptimeChecksConfigs(): Promise<GetUptimeChecksConfigResult>;

    listTimeSeries(filter: string, startTime: number, endTime: number): Promise<ListTimeSeriesResult>;
}

export async function getMonitoringUptimeCheckTimeSeries(
    googleMonitoring: GoogleMonitoring
): Promise<MonitoringDataModel[]> {
    const nowInSeconds = Date.now() / 1000;

    const startTime = nowInSeconds - 30;
    const endTime = nowInSeconds;

    try {
        const uptimeCheckConfigs = await googleMonitoring.getUptimeChecksConfigs();
        const monitoringDataResponse = uptimeCheckConfigs.map(fetchHostnames);
        return Promise.all(monitoringDataResponse);
    } catch (error: any) {
        console.error(`Response: ${error}`);
        return [{"hostname": "bts", "regions": [{"region": "europe", "status": "false"}]}];
    }

    async function fetchHostnames(uptimeCheckConfig: google.monitoring.v3.IUptimeCheckConfig): Promise<MonitoringDataModel> {
        const hostname = uptimeCheckConfig.monitoredResource?.labels?.host!;
        const regions = regionsMonitored.map((region) => fetchTimeSeriesPoints(region, hostname));
        return {
            hostname: hostname,
            regions: await Promise.all(regions)
        };
    }

    async function fetchTimeSeriesPoints(regionMonitored: string, hostname: string): Promise<Region> {
        const filter = "metric.type=\"monitoring.googleapis.com/uptime_check/check_passed\" resource.type=\"uptime_url\" " +
            `resource.label."host"="${hostname}" metric.label."checker_location"="${regionMonitored}"`;
        const status = await listTimeSeries(filter, hostname, regionMonitored);

        return {
            region: regionMonitored,
            status: status
        };
    }

    async function listTimeSeries(filter: string, hostname: string, regionMonitored: string): Promise<string> {
        try {
            const timeSeries = await googleMonitoring.listTimeSeries(filter, startTime, endTime);
            console.log(`Status: for ${hostname} in region = ${regionMonitored} is ${JSON.stringify(timeSeries[0].points?.at(0)?.value?.boolValue)}`);
            return timeSeries[0].points?.at(0)?.value?.boolValue == true ? "success" : "error";
        } catch (error: any) {
            console.error(`Response: ${error}`);
            console.log("Failed to get timeSeries points data");
            return "requestFailed";
        }
    }
}

//Google calling functions -- cannot be testes :)
// TODO: I would put this in a different file
export class GoogleMonitoringApi implements GoogleMonitoring {
    private uptimeClient: UptimeCheckServiceClient;
    private metricsClient: MetricServiceClient;

    constructor(public readonly projectId: string) {
        this.uptimeClient = new UptimeCheckServiceClient();
        this.metricsClient = new MetricServiceClient();
    }
    async getUptimeChecksConfigs(): Promise<GetUptimeChecksConfigResult> {

        const request = {
            parent: this.uptimeClient.projectPath(this.projectId)
        };

        // Retrieves an uptime check config with hostnames-services like dqs, bus, bum, dashboard etc
        const [uptimeCheckConfigs] = await this.uptimeClient.listUptimeCheckConfigs(request);

        return uptimeCheckConfigs;
    }

    async listTimeSeries(filter: string, startTime: number, endTime: number): Promise<ListTimeSeriesResult> {
        const request = {
            name: this.metricsClient.projectPath(this.projectId),
            filter: filter,
            interval: {
                startTime: {
                    // Limit results to the last 30 seconds
                    seconds: startTime,
                },
                endTime: {
                    seconds: endTime
                }
            }
        };
        // Writes time series data
        const [timeSeries] = await this.metricsClient.listTimeSeries(request);

        return timeSeries;
    }
}
