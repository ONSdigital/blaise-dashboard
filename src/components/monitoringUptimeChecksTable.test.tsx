/**
 * @jest-environment jsdom
 */

import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { mockHealthCheckList } from "../server/blaiseApi/testFixtures";

jest.mock("../client/monitoring");
import { getMonitoring } from "../client/monitoring";
import { MonitoringDataModel } from "../server/monitoringDataModel";

import MonitoringUptimeChecksTable from "./monitoringUptimeChecksTable";

const getMonitoringMock = getMonitoring as jest.Mock<Promise<MonitoringDataModel[]>>;


describe("Monitoring Uptime Checks Table", () => {
    

    afterEach(() => {
        jest.resetAllMocks();
    });

    it("renders a table row for the uptime checks", async () => {
        getMonitoringMock.mockImplementation(() => Promise.resolve(mockHealthCheckList));

        render(
            <MonitoringUptimeChecksTable monitoringData={mockHealthCheckList} />
        );

        expect(screen.queryByText(/Loading/i)).toBeDefined();

        await waitFor(() => {
            expect(screen.getByTestId("uptimecheck-dev-bts.social-surveys.gcp.onsdigital.uk").textContent).toEqual("dev-bts.social-surveys.gcp.onsdigital.uk");
            expect(screen.getByTestId("uptimecheck-europe").className).toContain("ons-status--success");
            expect(screen.getByTestId("uptimecheck-asia").className).toContain("ons-status--error");
            expect(screen.getByTestId("uptimecheck-northAmerica").className).toContain("ons-status--success");
            expect(screen.getByTestId("uptimecheck-southAmerica").className).toContain("ons-status--requestFailed");
            expect(screen.getByTestId("healthCheck-table")).toMatchSnapshot();
        });
    });

});
