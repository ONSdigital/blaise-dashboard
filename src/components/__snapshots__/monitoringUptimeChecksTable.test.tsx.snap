// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monitoring Uptime Checks Table renders a table row for the uptime checks 1`] = `
<table
  class="ons-table"
  data-testid="healthCheck-table"
  id="healthCheck-table"
>
  <thead
    class="ons-table__head"
  >
    <tr
      class="ons-table__row"
    >
      <th
        class="ons-table__header"
        scope="col"
      >
        Service
      </th>
      <th
        class="ons-table__header"
        scope="col"
      >
        Eur-Belgium
      </th>
      <th
        class="ons-table__header"
        scope="col"
      >
        Asia Pacific
      </th>
      <th
        class="ons-table__header"
        scope="col"
      >
        North America
      </th>
      <th
        class="ons-table__header"
        scope="col"
      >
        South America
      </th>
    </tr>
  </thead>
  <tbody
    class="ons-table__body"
  >
    <tr
      class="ons-table__row"
      data-testid="questionnaire-table-row"
    >
      <td
        class="ons-table__cell"
        data-testid="uptimecheck-dev-bts.social-surveys.gcp.onsdigital.uk"
      >
        dev-bts.social-surveys.gcp.onsdigital.uk
      </td>
      <td
        class="ons-table__cell"
      >
        <span
          class="ons-status ons-status--success"
          data-testid="uptimecheck-europe"
        />
      </td>
      <td
        class="ons-table__cell"
      >
        <span
          class="ons-status ons-status--error"
          data-testid="uptimecheck-asia"
        />
      </td>
      <td
        class="ons-table__cell"
      >
        <span
          class="ons-status ons-status--success"
          data-testid="uptimecheck-northAmerica"
        />
      </td>
      <td
        class="ons-table__cell"
      >
        <span
          class="ons-status ons-status--requestFailed"
          data-testid="uptimecheck-southAmerica"
        />
      </td>
    </tr>
  </tbody>
</table>
`;
